File: python/bullmq/queue.py
import asyncio
from typing import Union
from bullmq.event_emitter import EventEmitter
from bullmq.redis_connection import RedisConnection
from bullmq.types import QueueBaseOptions, RetryJobsOptions, JobOptions, PromoteJobsOptions
from bullmq.utils import extract_result, is_redis_cluster, get_cluster_nodes, get_node_client
from bullmq.scripts import Scripts
from bullmq.job import Job


class Queue(EventEmitter):
    """
    Instantiate a Queue object
    """

    def __init__(self, name: str, opts: QueueBaseOptions = {}):
        """
        Initialize a connection
        """
        self.name = name
        redisOpts = opts.get("connection", {})
        self.redisConnection = RedisConnection(redisOpts)
        self.client = self.redisConnection.conn
        self.opts = opts
        self.jobsOpts = opts.get("defaultJobOptions", {})
        self.prefix = opts.get("prefix", "bull")
        self.scripts = Scripts(
            self.prefix, name, self.redisConnection)
        self.keys = self.scripts.queue_keys.getKeys(name)
        self.qualifiedName = self.scripts.queue_keys.getQueueQualifiedName(name)

    def toKey(self, type: str):
        return self.scripts.queue_keys.toKey(self.name, type)

    async def add(self, name: str, data, opts: JobOptions = {}):
        """
        Adds a new job to the queue.

        @param name: Name of the job to be added to the queue,.
        @param data: Arbitrary data to append to the job.
        @param opts: Job options that affects how the job is going to be processed.
        """
        merged_opts = {**self.jobsOpts, **(opts or {})}
        job = Job(self, name, data, merged_opts)
        job_id = await self.scripts.addJob(job)
        job.id = job_id
        return job

    async def addBulk(self, jobs: list[dict[str, Union[dict, str]]]):
        """
        Adds an array of jobs to the queue. This method may be faster than adding
        one job at a time in a sequence
        """
        jobs_data = []
        for job in jobs:
            opts = {**self.jobsOpts, **(job.get("opts") or {})}

            jobs_data.append({
                "name": job.get("name"),
                "data": job.get("data"),
                "opts": opts
            })

        result = []
        async with self.redisConnection.conn.pipeline(transaction=True) as pipe:
            for job_data in jobs_data:
                current_job_opts = job_data.get("opts", {})
                job = Job(
                    queue=self,
                    name=job_data.get("name"),
                    data=job_data.get("data"),
                    opts=current_job_opts,
                    job_id=current_job_opts.get("jobId")
                    )
                job_id = await self.scripts.addJob(job, pipe)
                job.id = job_id
                result.append(job)
            job_ids = await pipe.execute()
            for index, job_id in enumerate(job_ids):
                result[index].id = job_id

        return result

    def pause(self):
        """
        Pauses the processing of this queue globally.

        We use an atomic RENAME operation on the wait queue. Since
        we have blocking calls with BRPOPLPUSH on the wait queue, as long as the queue
        is renamed to 'paused', no new jobs will be processed (the current ones
        will run until finalized).

        Adding jobs requires a LUA script to check first if the paused list exist
        and in that case it will add it there instead of the wait list.
        """
        return self.scripts.pause(True)

    def resume(self):
        """
        Resumes the processing of this queue globally.

        The method reverses the pause operation by resuming the processing of the
        queue.
        """
        return self.scripts.pause(False)

    async def isPaused(self):
        """
        Returns true if the queue is currently paused.
        """
        paused_key_exists = await self.client.hexists(self.keys["meta"], "paused")
        return paused_key_exists == 1

    def getRateLimitTtl(self):
        """
        Returns the time to live for a rate limited key in milliseconds.
        """
        return self.client.pttl(self.keys["limiter"])

    async def get_workers(self):
        """
        Get the worker list related to the queue. i.e. all the known
        workers that are available to process jobs for this queue.
        Note: Some Redis providers do not support CLIENT LIST.
        """
        client_name_prefix = self.qualifiedName

        def matcher(name: str):
            return name == client_name_prefix or name.startswith(f"{client_name_prefix}:w:")

        return await self._base_get_clients(matcher)

    async def get_workers_count(self):
        workers = await self.get_workers()
        return len(workers)

    async def _base_get_clients(self, matcher):
        client = self.client
        try:
            if is_redis_cluster(client):
                nodes = get_cluster_nodes(client)
                clients_per_node = []
                for node in nodes:
                    node_client = get_node_client(node)
                    client_list = await self._get_client_list(node_client)
                    clients_per_node.append(self._parse_client_list(client_list, matcher))

                if not clients_per_node:
                    return []

                return max(clients_per_node, key=len)

            client_list = await self._get_client_list(client)
            return self._parse_client_list(client_list, matcher)
        except Exception as err:
            message = str(err)
            if "unknown command" in message and "CLIENT" in message:
                return [{"name": "CLIENT LIST not supported"}]
            raise

    async def _get_client_list(self, client):
        if hasattr(client, "client_list"):
            return await client.client_list()
        return await client.execute_command("CLIENT", "LIST")

    def _parse_client_list(self, client_list, matcher):
        if isinstance(client_list, bytes):
            client_list = client_list.decode()

        clients = []
        if isinstance(client_list, list):
            clients = client_list
        elif isinstance(client_list, str):
            lines = client_list.splitlines()
            for line in lines:
                if not line:
                    continue
                client = {}
                for key_value in line.split(" "):
                    if "=" not in key_value:
                        continue
                    key, value = key_value.split("=", 1)
                    client[key] = value
                clients.append(client)

        result = []
        for client in clients:
            name = ""
            if isinstance(client, dict):
                name = client.get("name") or ""
                if isinstance(name, bytes):
                    name = name.decode()
            if matcher(name):
                client_copy = dict(client)
                client_copy["rawname"] = name
                client_copy["name"] = self.name
                result.append(client_copy)

        return result

    async def getJobLogs(self, job_id:str, start = 0, end = -1, asc = True):
        """
        Returns the logs for a given Job.

        @param job_id: The id of the job to get the logs for.
        @param start: Zero based index from where to start returning jobs.
        @param end: Zero based index where to stop returning jobs.
        @param asc: If true, the jobs will be returned in ascending order.
        """

        logs_key = self.toKey(job_id + ":logs")
        pipe = self.redisConnection.conn.pipeline(transaction=True)
        if asc:
            pipe.lrange(logs_key, start, end)
        else:
            pipe.lrange(logs_key, -(end+1), -(start+1))
        pipe.llen(logs_key)
        result = await pipe.execute()
        if not asc:
            result[0].reverse()
        return {
            "logs": result[0],
            "count": result[1]
        }
   
    async def obliterate(self, force: bool = False):
        """
        Completely destroys the queue and all of its contents irreversibly.
        This method will the *pause* the queue and requires that there are no
        active jobs. It is possible to bypass this requirement, i.e. not
        having active jobs using the "force" option.

        Note: This operation requires to iterate on all the jobs stored in the queue
        and can be slow for very large queues.

        @param opts: Obliterate options.
        """
        await self.pause()
        while True:
            cursor = await self.scripts.obliterate(1000, force)
            if cursor is None or cursor == 0 or cursor == "0":
                break

    async def drain(self, delayed: bool = False):
        """
        Drains the queue, removes all jobs that are waiting
        or delayed, but not active, completed or failed.
        
        @param delayed: Pass True if it should also clean the delayed jobs.
        """
        await self.scripts.drain(delayed)

    async def retryJobs(self, opts: RetryJobsOptions = {}):
        """
        Retry all the failed or completed jobs.
        """
        while True:
            cursor = await self.scripts.retryJobs(
                opts.get("state"),
                opts.get("count"),
                opts.get("timestamp")
            )
            if cursor is None or cursor == 0 or cursor == "0":
                break

    async def promoteJobs(self, opts: PromoteJobsOptions = {}):
        """
        Retry all the delayed jobs.
        """
        while True:
            cursor = await self.scripts.promoteJobs(
                opts.get("count")
            )
            if cursor is None or cursor == 0 or cursor == "0":
                break

    def trimEvents(self, maxLength: int):
        """
        Trim the event stream to an approximately maxLength.

        @param maxLength:
        """
        return self.client.xtrim(self.keys["events"], maxlen=maxLength, approximate="~")

    def removeDeprecatedPriorityKey(self):
        """
        Delete old priority helper key.
        """
        return self.client.delete(self.toKey("priority"))

    async def getJobCountByTypes(self, *types):
        result = await self.getJobCounts(*types)
        sum = 0
        for attribute in result:
            sum += result[attribute]
        return sum

    async def getJobCounts(self, *types):
        """
        Returns the job counts for each type specified or every list/set in the queue by default.

        @returns: An object, key (type) and value (count)
        """
        current_types = self.sanitizeJobTypes(types)

        responses = await self.scripts.getCounts(current_types)
        counts = {}

        for index, val in enumerate(responses):
            counts[current_types[index]] = val or 0
        return counts

    async def getCountsPerPriority(self, priorities):
        """
        Returns the number of jobs per priority.

        @returns: An object, key (priority) and value (count)
        """
        set_priorities = set(priorities)
        unique_priorities = (list(set_priorities))

        responses = await self.scripts.getCountsPerPriority(unique_priorities)

        counts = {}

        for index, val in enumerate(responses):
            counts[f"{unique_priorities[index]}"] = val or 0
        return counts

    async def clean(self, grace: int, limit: int, type: str):
        """
        Cleans jobs from a queue. Similar to drain but keeps jobs within a certain
        grace period
        
        * @returns: Id jobs from the deleted records
        """
        jobs = await self.scripts.cleanJobsInSet(type, grace, limit)

        return jobs

    def getJobState(self, job_id: str):
        return self.scripts.getState(job_id)

    def getCompletedCount(self):
        return self.getJobCountByTypes('completed')

    def getDelayedCount(self):
        return self.getJobCountByTypes('delayed')

    def getFailedCount(self):
        return self.getJobCountByTypes('failed')

    def getWaitingCount(self):
        return self.getJobCountByTypes('waiting')

    def getActive(self, start=0, end=-1):
        return self.getJobs(['active'], start, end, True)

    def getCompleted(self, start = 0, end=-1):
        return self.getJobs(['completed'], start, end, False)

    def getDelayed(self, start = 0, end=-1):
        return self.getJobs(['delayed'], start, end, True)

    def getFailed(self, start = 0, end=-1):
        return self.getJobs(['completed'], start, end, False)

    def getPrioritized(self, start = 0, end=-1):
        return self.getJobs(['prioritized'], start, end, True)

    def getWaiting(self, start = 0, end=-1):
        return self.getJobs(['waiting'], start, end, True)

    def getWaitingChildren(self, start = 0, end=-1):
        return self.getJobs(['waiting-children'], start, end, True)

    async def getJobs(self, types, start=0, end=-1, asc:bool=False):
        current_types = self.sanitizeJobTypes(types)
        job_ids = await self.scripts.getRanges(current_types, start, end, asc)
        tasks = [asyncio.create_task(Job.fromId(self, i)) for i in job_ids]
        job_set, _ = await asyncio.wait(tasks, return_when=asyncio.ALL_COMPLETED)
        jobs = [extract_result(job_task, self.emit) for job_task in job_set]
        jobs_len = len(jobs)

        # we filter `None` out to remove:
        jobs = list(filter(lambda j: j is not None, jobs))

        for index, job_id in enumerate(job_ids):
            pivot_job = jobs[index]

            for i in range(index,jobs_len):
                current_job = jobs[i]
                if current_job and current_job.id == job_id:
                    jobs[index] = current_job
                    jobs[i] = pivot_job
                    continue

        return jobs

    def sanitizeJobTypes(self, types):
        current_types = list(types)

        if len(types) > 0:
            sanitized_types = current_types.copy()

            try:
                sanitized_types.index('waiting')
                sanitized_types.append('paused')
            except ValueError:
                pass
            set_res = set(sanitized_types)
            list_res = (list(set_res))

            return list_res
        return [
            'active',
            'completed',
            'delayed',
            'failed',
            'paused',
            'waiting',
            'waiting-children'
        ]

    async def close(self):
        """
        Close the queue instance.
        """
        return await self.redisConnection.close()

    def remove(self, job_id: str, opts: dict = {}):
        return self.scripts.remove(job_id, opts.get("removeChildren", True))

File: php/src/Queue.php
<?php

declare(strict_types=1);

namespace BullMQ;

use Ramsey\Uuid\Uuid;

/**
 * Queue client for BullMQ.
 *
 * This class provides the ability to add jobs to a queue that can be processed
 * by workers written in NodeJS, Python, or Elixir.
 *
 * @example
 * ```php
 * use BullMQ\Queue;
 *
 * $queue = new Queue('my-queue');
 * $job = $queue->add('my-job', ['foo' => 'bar']);
 * ```
 */
class Queue
{
    /**
     * The name of the queue.
     */
    public readonly string $name;

    /**
     * The prefix used for Redis keys.
     */
    private string $prefix;

    /**
     * The Redis connection.
     */
    private RedisConnection $connection;

    /**
     * Scripts executor.
     */
    private Scripts $scripts;

    /**
     * Queue keys helper.
     */
    private QueueKeys $keys;

    /**
     * Whether the queue owns the connection (should close it on cleanup).
     */
    private bool $ownsConnection = false;

    /**
     * Create a new Queue instance.
     *
     * @param string $name The name of the queue
     * @param array{
     *   prefix?: string,
     *   connection?: RedisConnection|array<string, mixed>|string
     * } $opts Queue options
     */
    public function __construct(string $name, array $opts = [])
    {
        $this->name = $name;
        $this->prefix = $opts['prefix'] ?? 'bull';

        // Handle connection option
        $connection = $opts['connection'] ?? [];
        if ($connection instanceof RedisConnection) {
            $this->connection = $connection;
        } else {
            $this->connection = new RedisConnection($connection);
            $this->ownsConnection = true;
        }

        $this->keys = new QueueKeys($this->prefix);
        $this->scripts = new Scripts($this->prefix, $this->name, $this->connection);
    }

    /**
     * Get the Redis connection.
     */
    public function getConnection(): RedisConnection
    {
        return $this->connection;
    }

    /**
     * Get the qualified name of the queue.
     */
    public function getQualifiedName(): string
    {
        return $this->keys->getQueueQualifiedName($this->name);
    }

    /**
     * Adds a new job to the queue.
     *
     * @param string $name The name of the job
     * @param mixed $data The data payload for the job
     * @param JobOptions|array<string, mixed> $opts Job options
     * @return Job The created job
     *
     * @example
     * ```php
     * // Simple job
     * $job = $queue->add('email', ['to' => 'user@example.com']);
     *
     * // Delayed job
     * $job = $queue->add('email', $data, ['delay' => 60000]); // 60 seconds
     *
     * // Priority job
     * $job = $queue->add('email', $data, ['priority' => 1]);
     * ```
     */
    public function add(string $name, mixed $data, JobOptions|array $opts = []): Job
    {
        $job = new Job($this, $name, $data, $opts);

        // Generate job ID if not provided
        if ($job->id === null) {
            $job->id = $this->generateJobId();
        }

        $this->scripts->addJob($job);

        return $job;
    }

    /**
     * Adds multiple jobs to the queue in bulk.
     *
     * @param array<array{name: string, data: mixed, opts?: JobOptions|array<string, mixed>}> $jobs
     * @return array<Job> Array of created jobs
     *
     * @example
     * ```php
     * $jobs = $queue->addBulk([
     *     ['name' => 'email', 'data' => ['to' => 'user1@example.com']],
     *     ['name' => 'email', 'data' => ['to' => 'user2@example.com']],
     * ]);
     * ```
     */
    public function addBulk(array $jobs): array
    {
        $createdJobs = [];

        foreach ($jobs as $jobData) {
            $name = $jobData['name'];
            $data = $jobData['data'];
            $opts = $jobData['opts'] ?? [];

            $job = new Job($this, $name, $data, $opts);
            if ($job->id === null) {
                $job->id = $this->generateJobId();
            }
            $createdJobs[] = $job;
        }

        // Execute all jobs atomically using a transaction (MULTI/EXEC)
        $client = $this->connection->getClient();
        $transaction = $client->transaction();

        foreach ($createdJobs as $job) {
            $this->scripts->addJobToTransaction($transaction, $job);
        }

        $results = $transaction->execute();

        // Check results for errors
        foreach ($results as $index => $result) {
            if (is_int($result) && $result < 0) {
                throw new \RuntimeException("Failed to add job at index {$index}: error code {$result}");
            }
            // Update job ID from result if returned
            if (is_string($result) && !empty($result)) {
                $createdJobs[$index]->id = $result;
            }
        }

        return $createdJobs;
    }

    /**
     * Pauses the processing of the queue.
     *
     * @return void
     */
    public function pause(): void
    {
        $this->scripts->pause(true);
    }

    /**
     * Resumes the processing of the queue.
     *
     * @return void
     */
    public function resume(): void
    {
        $this->scripts->pause(false);
    }

    /**
     * Checks if the queue is paused.
     *
     * @return bool True if the queue is paused
     */
    public function isPaused(): bool
    {
        $client = $this->connection->getClient();
        $pausedKey = $this->keys->toKey($this->name, 'meta');
        
        $paused = $client->hget($pausedKey, 'paused');
        
        return $paused === '1';
    }

    /**
     * Get the total count of jobs in a specific state.
     *
     * @param string ...$types The job states to count (waiting, active, delayed, completed, failed, paused, prioritized, waiting-children)
     * @return int The total count
     */
    public function getJobCountByTypes(string ...$types): int
    {
        $counts = $this->scripts->getCounts($types);
        return array_sum($counts);
    }

    /**
     * Get job counts for all states.
     *
     * @return array{waiting: int, active: int, delayed: int, completed: int, failed: int, paused: int, prioritized: int, waiting-children: int}
     */
    public function getJobCounts(): array
    {
        $types = ['waiting', 'active', 'delayed', 'completed', 'failed', 'paused', 'prioritized', 'waiting-children'];
        $counts = $this->scripts->getCounts($types);

        $result = [];
        foreach ($types as $i => $type) {
            $result[$type] = $counts[$i] ?? 0;
        }

        return $result;
    }

    /**
     * Get counts per priority.
     *
     * @param array<int> $priorities Array of priority levels to count
     * @return array<int, int> Priority to count mapping
     */
    public function getCountsPerPriority(array $priorities): array
    {
        $counts = $this->scripts->getCountsPerPriority($priorities);
        
        $result = [];
        foreach ($priorities as $i => $priority) {
            $result[$priority] = $counts[$i] ?? 0;
        }

        return $result;
    }

    /**
     * Get a job by its ID.
     *
     * @param string $jobId The job ID
     * @return Job|null The job, or null if not found
     */
    public function getJob(string $jobId): ?Job
    {
        $client = $this->connection->getClient();
        $jobKey = $this->keys->toKey($this->name, $jobId);
        
        $rawData = $client->hgetall($jobKey);
        
        if (empty($rawData)) {
            return null;
        }

        return Job::fromRaw($this, $rawData, $jobId);
    }

    /**
     * Get jobs by their IDs.
     *
     * @param array<string> $jobIds The job IDs
     * @return array<Job> Array of jobs (excluding not found jobs)
     */
    public function getJobs(array $jobIds): array
    {
        $jobs = [];
        
        foreach ($jobIds as $jobId) {
            $job = $this->getJob($jobId);
            if ($job !== null) {
                $jobs[] = $job;
            }
        }

        return $jobs;
    }

    /**
     * Get jobs by type/state.
     *
     * @param string|array<string> $types Job states (waiting, active, delayed, completed, failed)
     * @param int $start Start index
     * @param int $end End index
     * @param bool $asc Sort order (true for ascending)
     * @return array<Job>
     */
    public function getJobsByType(string|array $types, int $start = 0, int $end = -1, bool $asc = false): array
    {
        if (is_string($types)) {
            $types = [$types];
        }

        $jobIds = $this->scripts->getRanges($types, $start, $end, $asc);
        
        return $this->getJobs($jobIds);
    }

    /**
     * Get waiting jobs.
     *
     * @param int $start Start index
     * @param int $end End index
     * @return array<Job>
     */
    public function getWaiting(int $start = 0, int $end = -1): array
    {
        return $this->getJobsByType('waiting', $start, $end);
    }

    /**
     * Get active jobs.
     *
     * @param int $start Start index
     * @param int $end End index
     * @return array<Job>
     */
    public function getActive(int $start = 0, int $end = -1): array
    {
        return $this->getJobsByType('active', $start, $end);
    }

    /**
     * Get delayed jobs.
     *
     * @param int $start Start index
     * @param int $end End index
     * @return array<Job>
     */
    public function getDelayed(int $start = 0, int $end = -1): array
    {
        return $this->getJobsByType('delayed', $start, $end);
    }

    /**
     * Get prioritized jobs.
     *
     * @param int $start Start index
     * @param int $end End index
     * @return array<Job>
     */
    public function getPrioritized(int $start = 0, int $end = -1): array
    {
        return $this->getJobsByType('prioritized', $start, $end);
    }

    /**
     * Get completed jobs.
     *
     * @param int $start Start index
     * @param int $end End index
     * @return array<Job>
     */
    public function getCompleted(int $start = 0, int $end = -1): array
    {
        return $this->getJobsByType('completed', $start, $end);
    }

    /**
     * Get failed jobs.
     *
     * @param int $start Start index
     * @param int $end End index
     * @return array<Job>
     */
    public function getFailed(int $start = 0, int $end = -1): array
    {
        return $this->getJobsByType('failed', $start, $end);
    }

    /**
     * Get the state of a job.
     *
     * @param string $jobId The job ID
     * @return string The job state (waiting, active, delayed, completed, failed, unknown)
     */
    public function getJobState(string $jobId): string
    {
        return $this->scripts->getState($jobId);
    }

    /**
     * Remove a job from the queue.
     *
     * @param string $jobId The job ID to remove
     * @param bool $removeChildren Whether to also remove child jobs
     * @return bool True if the job was removed
     */
    public function remove(string $jobId, bool $removeChildren = true): bool
    {
        $result = $this->scripts->remove($jobId, $removeChildren);
        return $result !== false && $result !== 0;
    }

    /**
     * Clean jobs from a set by state.
     *
     * @param int $grace Grace period in milliseconds
     * @param int $limit Maximum number of jobs to clean (0 for unlimited)
     * @param string $type Job state to clean (completed, failed, delayed, waiting, active)
     * @return array<string> Array of job IDs that were removed
     */
    public function clean(int $grace, int $limit = 0, string $type = 'completed'): array
    {
        return $this->scripts->cleanJobsInSet($type, $grace, $limit);
    }

    /**
     * Drain the queue (remove all waiting and delayed jobs).
     *
     * @param bool $delayed Whether to also drain delayed jobs
     * @return void
     */
    public function drain(bool $delayed = false): void
    {
        // We need to pause first, then clean
        $this->pause();
        
        try {
            // Clean all waiting jobs
            $this->clean(0, 0, 'waiting');
            
            if ($delayed) {
                $this->clean(0, 0, 'delayed');
            }
        } finally {
            $this->resume();
        }
    }

    /**
     * Obliterate (completely destroy) the queue.
     *
     * This removes ALL data associated with the queue including jobs, logs, etc.
     * Use with caution!
     *
     * @param array{force?: bool, count?: int} $opts Options
     *   - force: Force obliteration even with active jobs
     *   - count: Maximum jobs to remove per iteration (default 1000)
     * @return void
     */
    public function obliterate(array $opts = []): void
    {
        $force = $opts['force'] ?? false;
        $count = $opts['count'] ?? 1000;

        // First pause the queue
        $this->pause();

        // Keep iterating until fully obliterated
        while (true) {
            $cursor = $this->scripts->obliterate($count, $force);
            if ($cursor === 0) {
                break;
            }
        }
    }

    /**
     * Retry all failed or completed jobs.
     *
     * @param array{count?: int, state?: string, timestamp?: int} $opts Options:
     *   - count: Maximum number of jobs to retry per iteration (default: 1000)
     *   - state: 'failed' or 'completed' (default: 'failed')
     *   - timestamp: Only retry jobs before this timestamp in ms (default: now)
     * @return void
     */
    public function retryJobs(array $opts = []): void
    {
        $count = $opts['count'] ?? 1000;
        $state = $opts['state'] ?? 'failed';
        $timestamp = $opts['timestamp'] ?? (int)(microtime(true) * 1000);

        do {
            $cursor = $this->scripts->moveJobsToWait($state, $count, $timestamp);
        } while ($cursor > 0);
    }

    /**
     * Promote all delayed jobs to waiting.
     *
     * @param array{count?: int} $opts Options:
     *   - count: Maximum number of jobs to promote per iteration (default: 1000)
     * @return void
     */
    public function promoteJobs(array $opts = []): void
    {
        $count = $opts['count'] ?? 1000;
        // Use max timestamp to promote all delayed jobs regardless of scheduled time
        $maxTimestamp = PHP_INT_MAX;

        do {
            $cursor = $this->scripts->moveJobsToWait('delayed', $count, $maxTimestamp);
        } while ($cursor > 0);
    }

    /**
     * Check if the queue is maxed (rate limited).
     *
     * @return bool True if the queue is maxed
     */
    public function isMaxed(): bool
    {
        return $this->scripts->isMaxed();
    }

    /**
     * Get rate limit TTL.
     *
     * @param int|null $maxJobs Maximum jobs to consider
     * @return int TTL in milliseconds
     */
    public function getRateLimitTtl(?int $maxJobs = null): int
    {
        return $this->scripts->getRateLimitTtl($maxJobs);
    }

    /**
     * Close the queue connection.
     *
     * @return void
     */
    public function close(): void
    {
        if ($this->ownsConnection) {
            $this->connection->close();
        }
    }

    /**
     * Generate a unique job ID.
     */
    private function generateJobId(): string
    {
        return Uuid::uuid4()->toString();
    }

    /**
     * Destructor - clean up resources.
     */
    public function __destruct()
    {
        $this->close();
    }
}
